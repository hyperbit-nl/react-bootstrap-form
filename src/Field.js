import React from 'react';
import { Form } from 'react-bootstrap';
import { createContentLabel, createLabel, getIdLabel, getLayout, getSectionLabel, getSizeClass } from './FieldCommon';
import { getDatetimeFormat, getDatetimeString } from './Utils';

let Select;
let CreatableSelect;
let AsyncSelect;
let AsyncCreatableSelect;
let moment;
let Datetime;

try { Select = require('react-select').default; } catch (_) { Select = null; }
try { CreatableSelect = require('react-select/creatable').default; } catch (_) { CreatableSelect = null; }
try { AsyncSelect = require('react-select/async').default; } catch (_) { AsyncSelect = null; }
try { AsyncCreatableSelect = require('react-select/async-creatable').default; } catch (_) { AsyncCreatableSelect = null; }
try { moment = require('moment'); } catch (_) { moment = null; }
try { Datetime = require('react-datetime').default; } catch (_) { Datetime = null; }

let didWarnImportReactSelect = false;
let didWarnImportReactDatetime = false;

const getIdInput = ({name}) => (`rbf_${name}_input`);
const getIdRadio = (name, index) => (`rbf_${name}_${index}`);
const getIdHelp = ({name}) => (`rbf_${name}_help`);

const getNumber = (value) => (
  value === '' ? '' : +value
);

export const RBField = (props) => {
  const {type = 'input', name, as, hidden, isValid, isInvalid, help} = props;

  if (type === 'hidden') {
    return null;
  }

  let propsGroup = {controlId: name, as, hidden};
  const propsControl = {type, isValid, isInvalid}; // Removed name.

  let SectionHelp;

  if (help) {
    const idHelp = getIdHelp(props);
    SectionHelp = getSectionHelp(idHelp, help);
    propsControl['aria-describedby'] = idHelp;
  }

  const {Content, Label, shouldReturnNow} = getSectionControl(propsGroup, propsControl, props, SectionHelp, getSectionFeedback(props));

  if (!Content) {
    return null;
  }

  if (shouldReturnNow) {
    return Content;
  }

  return getLayout(propsGroup, Label, Content, props);
};

const getSectionHelp = (id, help) => {
  return (
    <Form.Text id={id} muted>
      {help}
    </Form.Text>
  );
};

const getSectionFeedback = ({error}) => {
  return (
    <Form.Control.Feedback type="invalid">
      {error}
    </Form.Control.Feedback>
  );
};

const getSectionControl = (propsGroup, propsControl, {type, ...props}, SectionHelp, SectionFeedback) => {
  if (type === 'radio') {
    const {name, value, size, label, hiddenLabel, disabled, options, inline, isValid, isInvalid, onChange, handleChange} = props;

    const propsLabelRadio = {
      className: getSizeClass(size),
    };

    let Radios;

    if (options) {
      Radios = options.map(({value: valueItem, label: labelItem}, index) => {
        const propsRadio = {
          name,
          type,
          checked: valueItem === value,
          onChange: () => {
            if (onChange) {
              onChange(valueItem);
            }

            handleChange(name, valueItem);
          },
          isValid,
          isInvalid,
          disabled,
        };

        const LabelRadio = createLabel(Form.Check.Label, propsLabelRadio, props, labelItem);

        return (
          <Form.Check key={index} type={type} id={getIdRadio(name, index)} inline={inline}>
            <Form.Check.Input {...propsRadio} />
            {LabelRadio}
          </Form.Check>
        );
      });

      propsGroup.role = 'radiogroup';
    }

    let Label;

    if (!hiddenLabel) {
      const propsLabel = {...props, className: 'rbf-group'};

      Label = getSectionLabel(propsLabel);
      propsGroup['aria-labelledby'] = getIdLabel(name);

    } else {
      propsGroup['aria-label'] = label;
    }

    const Content = (
      <>
        {Radios}
        {createControlHidden(isValid, isInvalid)}
        {SectionHelp}
        {SectionFeedback}
      </>
    );

    return {Content, Label};
  }

  if (type === 'checkbox' || type === 'switch') {
    const {name, value, size, label, hiddenLabel, disabled, inline, single, as, onChange, handleChange, hidden} = props;
    const shouldReturnNow = !single && !as;

    const propsCheck = {type, inline};
    const propsCheckClassNames = [];

    if (shouldReturnNow) {
      propsCheck.id = name;

      if (hidden) { // Setting the hidden prop regularly doesn't work.
        propsCheckClassNames.push('rbf-hidden');
      }
    }

    const propsCheckLabel = {
      id: getIdLabel(name),
      className: getSizeClass(size),
    };

    let CheckLabel;

    if (!hiddenLabel) {
      CheckLabel = createLabel(Form.Check.Label, propsCheckLabel, props);

    } else {
      if (type == 'switch') {
        CheckLabel = createLabel(Form.Check.Label, propsCheckLabel, props, <span className="sr-only">{label}</span>);

      } else {
        propsControl['aria-label'] = label;
      }

      propsCheckClassNames.push(getSizeClass(size, false));
    }

    if (propsCheckClassNames.length > 0) {
      propsCheck.className = propsCheckClassNames.join(' ');
    }

    propsControl.type = 'checkbox';
    propsControl.disabled = disabled;
    propsControl.checked = !!value;
    propsControl.onChange = () => {
      const valueNew = !value;

      if (onChange) {
        onChange(valueNew);
      }

      handleChange(name, valueNew);
    };

    const Content = (
      <Form.Check {...propsCheck} hidden>
        <Form.Check.Input {...propsControl} />
        {CheckLabel}
        {SectionHelp}
        {SectionFeedback}
      </Form.Check>
    );

    return {Content, shouldReturnNow};
  }

  if (type === 'datetime') {
    const errorDatetime = [];

    if (!moment) {
      errorDatetime.push('\'moment\'');
    }

    if (!Datetime) {
      errorDatetime.push('\'react-datetime\'');
    }

    if (errorDatetime.length > 0) {
      if (!didWarnImportReactDatetime) {
        console.warn(`To use advanced datetime input, please install ${errorDatetime.join(' and ')}`);
        didWarnImportReactDatetime = true;
      }

      return null;
    }

    const {name, value, label, hiddenLabel, placeholder, disabled, readOnly, plaintext, size, dateFormat, timeFormat, locale, onChange, handleChange, isValid, isInvalid} = props;
    const datetimeFormat = getDatetimeFormat(dateFormat, timeFormat);

    const propsControl = {
      dateFormat,
      timeFormat,
      locale,
      initialValue: value,
      onChange: (value) => {
        const valueString = getDatetimeString(value, datetimeFormat, locale);

        if (onChange) {
          onChange(valueString);
        }

        handleChange(name, valueString);
      },
      closeOnSelect: true,
    };

    const inputProps = {
      placeholder,
      disabled,
      readOnly,
      className: 'form-control',
    };

    let Label;

    if (!hiddenLabel) {
      Label = getSectionLabel(props);
      inputProps['aria-labelledby'] = getIdLabel(name);

    } else {
      inputProps['aria-label'] = label;
    }

    if (readOnly) {
      propsControl.open = false;
    }

    if (!value) { // Needed to force reset.
      inputProps.value = '';
    }

    if (plaintext) {
      inputProps.className += ' form-control-plaintext';
    }

    if (size) {
      if (size === 'sm') {
        inputProps.className += ' form-control-sm';

      } else if (size === 'lg') {
        inputProps.className += ' form-control-lg';
      }
    }

    propsControl.inputProps = inputProps;

    const Content = (
      <>
        <Datetime {...propsControl} />
        {createControlHidden(isValid, isInvalid)}
        {SectionHelp}
        {SectionFeedback}
      </>
    );

    return {Content, Label};
  }

  if (type === 'select') {
    const {options} = props;

    if (options) {
      if (!Select) {
        if (!didWarnImportReactSelect) {
          console.warn('To use advanced select input, please install \'react-select\'');
          didWarnImportReactSelect = true;
        }

        return null;
      }

      const {name, value, label, hiddenLabel, placeholder, disabled, isSearchable, isClearable, isMulti, isCreatable, noOptionsMessage, async, defaultOptions, cacheOptions, isValid, isInvalid, onChange, handleChange} = props;

      const propsControl = {
        inputId: getIdInput(props),
        placeholder,
        isDisabled: disabled,
        isClearable,
        isSearchable,
        isMulti,
        noOptionsMessage,
        value: value ?? null, // Needed to force reset.
        onChange: (option) => {
          handleChange(name, option);

          if (onChange) {
            onChange(option);
          }
        },
        classNamePrefix: isInvalid ? 'react-select-invalid' : isValid ? 'react-select-valid': 'react-select',
      };

      if (disabled) {
        propsControl.styles = {
          control: (styles) => ({
            ...styles,
            backgroundColor: '#e9ecef',
          })
        };
      }

      let SelectType;

      if (async) {
        propsControl.loadOptions = options;
        propsControl.defaultOptions = defaultOptions;
        propsControl.cacheOptions = cacheOptions;

        SelectType = isCreatable ? AsyncCreatableSelect : AsyncSelect;

      } else {
        propsControl.options = options;

        SelectType = isCreatable ? CreatableSelect : Select;
      }

      let Label;

      if (!hiddenLabel) {
        Label = getSectionLabel(props);
        propsControl['aria-labelledby'] = getIdLabel(name);

      } else {
        propsControl['aria-label'] = label;
      }

      const Content = (
        <>
          <SelectType {...propsControl} />
          {createControlHidden(isValid, isInvalid)}
          {SectionHelp}
          {SectionFeedback}
        </>
      );

      return {Content, Label};
    }

    const {name, value, plaintext, size, disabled, multiple, htmlSize, children, label, hiddenLabel, onChange, handleChange} = props;
    let Label;

    if (!hiddenLabel) {
      Label = getSectionLabel(props);

    } else {
      propsControl['aria-label'] = label;
    }

    propsControl.plaintext = plaintext;
    propsControl.disabled = disabled;
    propsControl.size = size;
    propsControl.as = type;
    propsControl.children = children;
    propsControl.value = value;

    if (multiple || htmlSize) {
      propsControl.multiple = true;
      propsControl.htmlSize = htmlSize;
      propsControl.onChange = (e) => {
        const value = [].slice.call(e.target.selectedOptions).map((option) => (option.value));

        if (onChange) {
          onChange(value);
        }

        handleChange(name, value);
      };

    } else {
      propsControl.onChange = (e) => {
        const value = e.target.value;

        if (onChange) {
          onChange(value);
        }

        handleChange(name, value);
      };
    }

    const Content = (
      <>
        <Form.Control {...propsControl} />
        {SectionHelp}
        {SectionFeedback}
      </>
    );

    return {Content, Label};
  }

  if (type === 'file') {
    const {name, value, disabled, accept, multiple, ['data-browse']: dataBrowse, label, hiddenLabel, onChange, handleChange} = props;

    propsControl.disabled = disabled;
    propsControl.onChange = (e) => {
      const value = e.target.files;

      if (onChange) {
        onChange(value);
      }

      handleChange(name, value);
    };
    propsControl.accept = accept;
    propsControl.multiple = multiple;

    let SelectedFiles;

    if (value) {
      propsControl.files = value;

      SelectedFiles = (
        <span muted>
          {Array.from(value).map(({name}) => (name)).join(', ')}
        </span>
      );
    }

    const Label = createLabel(Form.File.Label, {['data-browse']: dataBrowse}, props, !hiddenLabel ? createContentLabel(props) : <span className="sr-only">{label}</span>);

    const Content = (
      <Form.File custom>
        <Form.File.Input {...propsControl} />
        {Label}
        {SelectedFiles}
        {SectionHelp}
        {SectionFeedback}
      </Form.File>
    );

    return {Content};
  }

  if (type === 'number' || type === 'range') {
    const {name, value, max, min, step, plaintext, readOnly, disabled, size, placeholder, label, hiddenLabel, onChange, handleChange} = props;

    let Label;

    if (!hiddenLabel) {
      Label = getSectionLabel(props);

    } else {
      propsControl['aria-label'] = label;
    }

    if (type === 'number') {
      propsControl.placeholder = placeholder;
      propsControl.size = size;
      propsControl.plaintext = plaintext;
      propsControl.readOnly = readOnly;
    }

    propsControl.disabled = disabled;
    propsControl.max = max;
    propsControl.min = min;
    propsControl.step = step;

    propsControl.value = getNumber(value);
    propsControl.onChange = (e) => {
      const value = getNumber(e.target.value);

      if (onChange) {
        onChange(value);
      }

      handleChange(name, value);
    };

    const Content = (
      <>
        <Form.Control {...propsControl} />
        {SectionHelp}
        {SectionFeedback}
      </>
    );

    return {Content, Label};
  }

  if (type === 'textarea') {
    const {name, value, plaintext, readOnly, disabled, size, placeholder, rows, label, hiddenLabel, onChange, handleChange} = props;

    let Label;

    if (!hiddenLabel) {
      Label = getSectionLabel(props);

    } else {
      propsControl['aria-label'] = label;
    }

    propsControl.as = type;
    propsControl.rows = rows;
    propsControl.placeholder = placeholder;
    propsControl.size = size;
    propsControl.plaintext = plaintext;
    propsControl.readOnly = readOnly;
    propsControl.disabled = disabled;
    propsControl.value = value;
    propsControl.onChange = (e) => {
      const value = e.target.value;

      if (onChange) {
        onChange(value);
      }

      handleChange(name, value);
    };

    const Content = (
      <>
        <Form.Control {...propsControl} />
        {SectionHelp}
        {SectionFeedback}
      </>
    );

    return {Content, Label};
  }

  const {name, value, plaintext, readOnly, disabled, size, placeholder, label, hiddenLabel, onChange, handleChange} = props;

  let Label;

  if (!hiddenLabel) {
    Label = getSectionLabel(props);

  } else {
    propsControl['aria-label'] = label;
  }

  propsControl.placeholder = placeholder;
  propsControl.size = size;
  propsControl.plaintext = plaintext;
  propsControl.readOnly = readOnly;
  propsControl.disabled = disabled;
  propsControl.value = value;
  propsControl.onChange = (e) => {
    const value = e.target.value;

    if (onChange) {
      onChange(value);
    }

    handleChange(name, value);
  };

  const Content = (
    <>
      <Form.Control {...propsControl} />
      {SectionHelp}
      {SectionFeedback}
    </>
  );

  return {Content, Label};
};

const createControlHidden = (isValid, isInvalid) => (
  <Form.Control {...{type: 'hidden', isValid, isInvalid}} />
);
