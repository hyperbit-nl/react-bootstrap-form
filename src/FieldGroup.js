import { getIdLabel, getLayout, getSectionLabel } from './FieldCommon';

export const RBFieldGroup = (props) => {
  const {id, label, hiddenLabel, as, children} = props;

  const propsGroup = {id, role: 'group', as};
  const propsLabel = {...props, name: id, className: 'rbf-group'};

  let Label;

  if (!hiddenLabel) {
    Label = getSectionLabel(propsLabel);
    propsGroup['aria-labelledby'] = getIdLabel(id);

  } else {
    propsGroup['aria-label'] = label;
  }

  return getLayout(propsGroup, Label, children, props);
};
