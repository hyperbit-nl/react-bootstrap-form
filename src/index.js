export {
  RBForm as default
} from './Form';

export {
  RBFieldGroup as FieldGroup
} from './FieldGroup';

export {
  RBFieldArray as FieldArray
} from './FieldArray';

export {
  RBField as Field
} from './Field';
