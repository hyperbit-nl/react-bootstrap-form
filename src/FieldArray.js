import React from 'react';
import { FieldArray } from 'formik';
import { Button } from 'react-bootstrap';
import { getLayout, getSectionLabel, RBArrayAssignProps, RBCustomAssignProps, RBFieldAssignProps, RBFormObjectRecursiveAssignProps } from './FieldCommon';
import { RBField } from './Field';

export const RBFieldArray = (props) => {
  const {name, ...rest} = props; // Make sure to remove 'name' form the rest.
  const {values, errors, didSubmbit, handleChange, initialValues, addButton, noRowsMessage, children, as, label} = rest;

  const Component = (
    <FieldArray {...{name}} render={({push, remove}) => {
      const addFunction = () => (push(initialValues));

      const Add = (addButton === undefined || addButton == true) ? (
        <Button className="mb-3" variant="primary" type="button" onClick={addFunction}>Add</Button>
      ) : (
        !addButton ? null : addButton(addFunction)
      );

      return (
        <>
          {values && values.length > 0 ? (
            values.map((_, indexValue) => (
              <React.Fragment key={indexValue}>
                {RBFormObjectRecursiveAssignProps(children, (child) => {
                  const {type, props} = child;
                  const {type: typeValue, nameValue, customInput} = props;
                  const typeComponent = customInput ? 'custom' : type;

                  switch (typeComponent) {
                    case RBField:
                      return RBFieldAssignProps({
                        name: nameValue,
                        type: typeValue,
                        values,
                        errors,
                        didSubmbit,
                        handleChange,
                        nameArray: name,
                        indexValue,
                      }, true);

                    case RBFieldArray:
                      return RBArrayAssignProps({
                        name: nameValue,
                        values,
                        errors,
                        didSubmbit,
                        handleChange,
                        initialValues,
                        nameArray: name,
                        indexValue,
                      }, true);

                    case 'custom':
                      return RBCustomAssignProps({
                        name: nameValue,
                        type: typeValue,
                        values,
                        errors,
                        didSubmbit,
                        handleChange,
                        nameArray: name,
                        indexValue,
                      }, true);

                    default: {
                      if (props['array-remove']) {
                        return {
                          ['array-remove']: 'true',
                          onClick: () => {
                            remove(indexValue);
                          },
                        };
                      }
                      return;
                    }
                  }
                })}
              </React.Fragment>
            ))
          ) : (
            noRowsMessage
          )}
          {Add}
        </>
      );
    }} />
  );

  return getLayout({as}, label && getSectionLabel(rest), Component, rest);
};
