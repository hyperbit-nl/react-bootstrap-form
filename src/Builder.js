import React, { useEffect } from 'react';
import { Button, Form, Spinner } from 'react-bootstrap';
import { RBArrayAssignProps, RBCustomAssignProps, RBFieldAssignProps, RBFormObjectRecursiveAssignProps } from './FieldCommon';
import { RBField } from './Field';
import { RBFieldArray } from './FieldArray';

const createSubmitSpinner = () => (
  <Spinner className="ml-2 align-middle" animation="border" variant="primary" role="status">
    <span className="sr-only">Waiting for response</span>
  </Spinner>
);

const createSubmit = (sendText, {isLoadingSubmit, errorSubmit, successSubmit, didSubmbit, isValid, validationErrorText}) => (
  <div>
    <Button disabled={isLoadingSubmit} variant="primary" type="submit">{sendText}</Button>
    {isLoadingSubmit && createSubmitSpinner()}
    {!errorSubmit && !successSubmit && didSubmbit && !isValid ? (
      <span className="ml-2 rbf-text-red" role="status">{validationErrorText}</span>
    ) : !isLoadingSubmit ? (
      errorSubmit && (
        <span className="ml-2 rbf-text-red" role="status">{errorSubmit}</span>
      )
    ) ?? (
      successSubmit && (
        <span className="ml-2 rbf-text-green" role="status">{successSubmit}</span>
      )
    ) : null}
  </div>
);

const getSubmit = (customSubmit, props) => {
  if (customSubmit) {
    if (typeof customSubmit === 'string') {
      return createSubmit(customSubmit, props);
    }

    const {isLoadingSubmit, successSubmit, errorSubmit} = props;
    return customSubmit(isLoadingSubmit, successSubmit, errorSubmit);
  }

  return createSubmit('Send', props);
};

export const RBBuilder = ({
  form: {
    validateForm,
    handleSubmit,
    values,
    setFieldValue: handleChange,
    errors,
    setErrors,
    submitCount,
    isValid
  },
  validationSchema,
  initialValues,
  customSubmit,
  isLoadingSubmit,
  errorSubmit,
  fieldErrorSubmit,
  successSubmit,
  validationErrorText = 'Error',
  children
}) => {
  const didSubmbit = submitCount > 0;

  useEffect(() => {
    if (didSubmbit) {
      validateForm();
    }
  }, [didSubmbit, validateForm, validationSchema]);

  useEffect(() => {
    if (fieldErrorSubmit) {
      setErrors(fieldErrorSubmit);
    }
  });

  return (
    <Form
      noValidate
      onSubmit={handleSubmit}
    >
      {RBFormObjectRecursiveAssignProps(children, ({type, props}) => {
        const typeComponent = props.customInput ? 'custom' : type;

        switch (typeComponent) {
          case RBField:
            return RBFieldAssignProps({...props, values, errors, didSubmbit, handleChange}, false);

          case RBFieldArray:
            return RBArrayAssignProps({...props, values, errors, didSubmbit, handleChange, initialValues}, false);

          case 'custom':
            return RBCustomAssignProps({...props, values, errors, didSubmbit, handleChange}, false);

          default:
            return;
        }
      })}
      {getSubmit(customSubmit, {isLoadingSubmit, errorSubmit, successSubmit, didSubmbit, isValid, validationErrorText})}
    </Form>
  );
};
