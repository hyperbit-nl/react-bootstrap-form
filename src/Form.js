import React, { useEffect, useReducer } from 'react';
import { Formik } from 'formik';
import { RBFormObjectRecursiveForEach } from './Setup';
import { RBBuilder } from './Builder';
import { getShapeObject } from './Validation';
import { isArray, isObject, usePrevious } from './Utils';

const filterEditable = (target, source) => {
  if (isObject(target) && isObject(source)) {
    target = {...target};

    for (const key in source) {
      if (source[key] === true) { // True means that key must be hidden, otherwise this is undefined.
        delete target[key];

      } else if (isObject(target[key])) {
        target[key] = filterEditable(target[key], source[key]);

      } else if (isArray(target[key]) && Object.prototype.hasOwnProperty.call(source, key)) {
        target[key] = target[key].map((value) => (
          filterEditable(value, source[key])
        ));
      }
    }
  }

  return target;
};

const initialStateForm = {
  didLoad: false,
  validationSchema: null,
  valuesHidden: null,
  initialValues: null,
  values: null,
  isLoadingSubmit: false,
  successSubmit: false,
  errorSubmit: null,
  fieldErrorSubmit: null,
};

const reducerForm = (state, action) => {
  switch (action.type) {
    case 'update': {
      const {validationShape, values, didChangeInitialValues, ...rest} = action.payload;

      const stateNew = {
        didLoad: true,
        validationSchema: getShapeObject(validationShape),
        ...rest,
      };

      if (didChangeInitialValues) {
        stateNew.values = values;
      }

      return {...state, ...stateNew};
    }

    case 'startSubmit': {
      return {...state, isLoadingSubmit: true};
    }

    case 'submitSuccess': {
      const response = action.payload;

      return {
        ...state,
        isLoadingSubmit: false,
        successSubmit: response,
        errorSubmit: null,
        fieldErrorSubmit: null,
      };
    }

    case 'submitFailure': {
      const error = action.payload;
      const stateNew = {isLoadingSubmit: false, successSubmit: false};

      if (error) {
        const {server: serverError, field: fieldErrors} = action.payload;

        if (fieldErrors) {
          stateNew.fieldErrorSubmit = fieldErrors;
        }

        stateNew.errorSubmit = serverError ?? 'Error';

      } else {
        stateNew.errorSubmit = 'Error';
      }

      return {...state, ...stateNew};
    }

    default:
      break;
  }
};

export const RBForm = ({
  initialValues: initialValuesForm = false,
  validationErrorText,
  customSubmit,
  onSubmit,
  children
}) => {
  const initialValuesFormPrevious = usePrevious(initialValuesForm);

  const [state, dispatch] = useReducer(reducerForm, initialStateForm);

  useEffect(() => {
    const validationShape = {};
    const valuesHidden = {};
    const initialValues = {};
    const values = {};

    RBFormObjectRecursiveForEach(
      validationShape,
      valuesHidden,
      initialValues,
      isObject(initialValuesForm) ? initialValuesForm : {},
      values,
      isArray(children) ? children: [children]
    );

    dispatch({
      type: 'update',
      payload: {
        validationShape,
        valuesHidden,
        initialValues,
        values,
        didChangeInitialValues: initialValuesFormPrevious !== initialValuesForm,
      },
    });

  }, [initialValuesForm, initialValuesFormPrevious, children]);

  const handleSubmit = (values, {resetForm}) => {
    const valuesEditable = filterEditable(values, state.valuesHidden);

    if (onSubmit) {
      dispatch({type: 'startSubmit'});

      onSubmit(valuesEditable, (response) => {
        onSubmitSuccess(resetForm, response);

      }, (error) => {
        dispatch({type: 'submitFailure', payload: error});
      });

    } else {
      onSubmitSuccess(resetForm);
    }
  };

  const onSubmitSuccess = (resetForm, response = 'Success') => {
    resetForm();
    dispatch({type: 'submitSuccess', payload: response});
  };

  return (
    state.didLoad && (
      <Formik
        validationSchema={state.validationSchema}
        initialValues={state.values}
        onSubmit={handleSubmit}
        enableReinitialize
      >
        {(form) => (
          <RBBuilder
            {...{
              form,
              validationSchema: state.validationSchema,
              initialValues: state.initialValues,
              customSubmit,
              isLoadingSubmit: state.isLoadingSubmit,
              errorSubmit: state.errorSubmit,
              fieldErrorSubmit: state.fieldErrorSubmit,
              successSubmit: state.successSubmit,
              validationErrorText,
            }}
          >
            {children}
          </RBBuilder>
        )}
      </Formik>
    )
  );
};
