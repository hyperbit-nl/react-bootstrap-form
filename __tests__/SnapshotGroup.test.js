import React from 'react';
import { render } from '@testing-library/react';
import { RBForm } from '../src/Form';
import { RBFieldGroup as FieldGroup  } from '../src/FieldGroup';
import { RBField as Field } from '../src/Field';

it("renders default Group", () => {
  const {container} = render(
    <RBForm>
      <FieldGroup id="group" label="Group" />
    </RBForm>
  );

  expect(container).toMatchSnapshot();
});

it("renders default Group with hiddenLabel", () => {
  const {container} = render(
    <RBForm>
      <FieldGroup id="group" label="Group" hiddenLabel />
    </RBForm>
  );

  expect(container).toMatchSnapshot();
});

it("renders Group with one text Field", () => {
  const {container} = render(
    <RBForm>
      <FieldGroup id="group" label="Group">
        <Field name="text" type="text" label="Text" />
      </FieldGroup>
    </RBForm>
  );

  expect(container).toMatchSnapshot();
});

it("renders Group with two text Fields", () => {
  const {container} = render(
    <RBForm>
      <FieldGroup id="group" label="Group">
        <Field name="text_1" type="text" label="Text 1" />
        <Field name="text_2" type="text" label="Text 2" />
      </FieldGroup>
    </RBForm>
  );

  expect(container).toMatchSnapshot();
});

it("renders Group with one checkbox Field", () => {
  const {container} = render(
    <RBForm>
      <FieldGroup id="group" label="Group">
        <Field name="checkbox" type="checkbox" label="Checkbox" />
      </FieldGroup>
    </RBForm>
  );

  expect(container).toMatchSnapshot();
});

it("renders Group with two checkbox Fields", () => {
  const {container} = render(
    <RBForm>
      <FieldGroup id="group" label="Group">
        <Field name="checkbox_1" type="checkbox" label="Checkbox 1" />
        <Field name="checkbox_2" type="checkbox" label="Checkbox 2" />
      </FieldGroup>
    </RBForm>
  );

  expect(container).toMatchSnapshot();
});
