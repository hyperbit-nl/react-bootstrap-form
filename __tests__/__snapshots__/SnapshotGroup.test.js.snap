// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Group with one checkbox Field 1`] = `
<div>
  <form
    class=""
    novalidate=""
  >
    <div
      aria-labelledby="rbf_group_label"
      class="form-group"
      id="group"
      role="group"
    >
      <label
        class="rbf-group  form-label"
        id="rbf_group_label"
      >
        Group
      </label>
      <div
        class="form-check"
      >
        <input
          class="form-check-input"
          id="checkbox"
          type="checkbox"
        />
        <label
          class="form-check-label"
          for="checkbox"
          id="rbf_checkbox_label"
        >
          Checkbox
        </label>
        <div
          class="invalid-feedback"
        />
      </div>
    </div>
    <div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Send
      </button>
    </div>
  </form>
</div>
`;

exports[`renders Group with one text Field 1`] = `
<div>
  <form
    class=""
    novalidate=""
  >
    <div
      aria-labelledby="rbf_group_label"
      class="form-group"
      id="group"
      role="group"
    >
      <label
        class="rbf-group  form-label"
        id="rbf_group_label"
      >
        Group
      </label>
      <div
        class="form-group"
      >
        <label
          class="form-label"
          for="text"
          id="rbf_text_label"
        >
          Text
        </label>
        <input
          class="form-control"
          id="text"
          type="text"
          value=""
        />
        <div
          class="invalid-feedback"
        />
      </div>
    </div>
    <div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Send
      </button>
    </div>
  </form>
</div>
`;

exports[`renders Group with two checkbox Fields 1`] = `
<div>
  <form
    class=""
    novalidate=""
  >
    <div
      aria-labelledby="rbf_group_label"
      class="form-group"
      id="group"
      role="group"
    >
      <label
        class="rbf-group  form-label"
        id="rbf_group_label"
      >
        Group
      </label>
      <div
        class="form-check"
      >
        <input
          class="form-check-input"
          id="checkbox_1"
          type="checkbox"
        />
        <label
          class="form-check-label"
          for="checkbox_1"
          id="rbf_checkbox_1_label"
        >
          Checkbox 1
        </label>
        <div
          class="invalid-feedback"
        />
      </div>
      <div
        class="form-check"
      >
        <input
          class="form-check-input"
          id="checkbox_2"
          type="checkbox"
        />
        <label
          class="form-check-label"
          for="checkbox_2"
          id="rbf_checkbox_2_label"
        >
          Checkbox 2
        </label>
        <div
          class="invalid-feedback"
        />
      </div>
    </div>
    <div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Send
      </button>
    </div>
  </form>
</div>
`;

exports[`renders Group with two text Fields 1`] = `
<div>
  <form
    class=""
    novalidate=""
  >
    <div
      aria-labelledby="rbf_group_label"
      class="form-group"
      id="group"
      role="group"
    >
      <label
        class="rbf-group  form-label"
        id="rbf_group_label"
      >
        Group
      </label>
      <div
        class="form-group"
      >
        <label
          class="form-label"
          for="text_1"
          id="rbf_text_1_label"
        >
          Text 1
        </label>
        <input
          class="form-control"
          id="text_1"
          type="text"
          value=""
        />
        <div
          class="invalid-feedback"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class="form-label"
          for="text_2"
          id="rbf_text_2_label"
        >
          Text 2
        </label>
        <input
          class="form-control"
          id="text_2"
          type="text"
          value=""
        />
        <div
          class="invalid-feedback"
        />
      </div>
    </div>
    <div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Send
      </button>
    </div>
  </form>
</div>
`;

exports[`renders default Group 1`] = `
<div>
  <form
    class=""
    novalidate=""
  >
    <div
      aria-labelledby="rbf_group_label"
      class="form-group"
      id="group"
      role="group"
    >
      <label
        class="rbf-group  form-label"
        id="rbf_group_label"
      >
        Group
      </label>
    </div>
    <div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Send
      </button>
    </div>
  </form>
</div>
`;

exports[`renders default Group with hiddenLabel 1`] = `
<div>
  <form
    class=""
    novalidate=""
  >
    <div
      aria-label="Group"
      class="form-group"
      id="group"
      role="group"
    />
    <div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Send
      </button>
    </div>
  </form>
</div>
`;
